File: examples/dmd_binary_testimages/csharp/dmd_binary_testimages.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace dmd_binary_testimages_namespace
{
    class dmd_binary_testimages
    {
        // creates an Ajile project and returns in
        static Project CreateProject(ushort sequenceID = 1, uint sequenceRepeatCount = 0, float frameTime_ms = -1, ComponentList components = null)
        {

            string projectName = "dmd_binary_testimages_example";
            string filenameBase = "../../images/cat_";
            if (frameTime_ms < 0)
                frameTime_ms = 100;
            int numImages = 14;
            string filename;

            // create a new project
            Project project = new Project(projectName);
            if (components != null && components.Count > 0)
                project.SetComponents(components);

            // create the images
            for (ushort i = 1; i <= numImages; i++)
            {
                filename = filenameBase + i.ToString() + ".png";
                Image testImage = new Image(i);
                testImage.ReadFromFile(filename, DeviceType_e.DMD_4500_DEVICE_TYPE);
                project.AddImage(testImage);
            }

            // create the sequence
            project.AddSequence(new Sequence(sequenceID, projectName, DeviceType_e.DMD_4500_DEVICE_TYPE, SequenceType_e.SEQ_TYPE_PRELOAD, sequenceRepeatCount));

            // create a single sequence item, which all the frames will be added to
            project.AddSequenceItem(new SequenceItem(sequenceID, 1));

            // create the frames and add them to the project, which adds them to the last sequence item
            for (ushort i = 1; i <= numImages; i++)
            {
                Frame frame = new Frame();
                frame.SetSequenceID(sequenceID);
                frame.SetImageID(i);
                frame.SetFrameTimeMSec(frameTime_ms);
                project.AddFrame(frame);
            }

            return project;
        }

        static int RunExample(string[] args)
        {

            // default connection settings
            string ipAddress = "192.168.200.1";
            string netmask = "255.255.255.0";
            string gateway = "0.0.0.0";
            ushort port = 5005;
            CommunicationInterfaceType_e commInterface = CommunicationInterfaceType_e.USB2_INTERFACE_TYPE;

            // default sequence settings
            uint repeatCount = 0; // repeat forever
            float frameTime_ms = -1.0f; // frame time in milliseconds
            ushort sequenceID = 1;

            // read command line arguments
            for (int i = 1; i < args.Length; i++)
            {
                if (args[i] == "-i")
                {
                    ipAddress = args[++i];
                }
                else if (args[i] == "-r")
                {
                    repeatCount = ushort.Parse(args[++i]);
                }
                else if (args[i] == "-f")
                {
                    frameTime_ms = float.Parse(args[++i]);
                }
                else if (args[i] == "--usb3")
                {
                    commInterface = CommunicationInterfaceType_e.USB3_INTERFACE_TYPE;
                }
                else
                {
                    Console.WriteLine("Usage: " + args[0] + " [-i <IP address>] [-r <repeat count>] [-f <frame rate in ms>] [--usb3]\n");
                    return -1;
                }
            }

            // connect to the device
            HostSystem ajileSystem = new HostSystem();
            ajileSystem.SetConnectionSettingsStr(ipAddress, netmask, gateway, port);
            ajileSystem.SetCommunicationInterface(commInterface);
            if (ajileSystem.StartSystem() != ErrorType_e.ERROR_NONE)
            {
                Console.WriteLine("Error starting AjileSystem.");
                return -1;
            }

            // create the project
            Project project = CreateProject(sequenceID, repeatCount, frameTime_ms, ajileSystem.GetProject().Components());

            // get the first valid component index which will run the sequence
            bool wasFound = false;
            Sequence sequence = project.FindSequence(sequenceID, out wasFound);
            if (!wasFound) return -1;
            int componentIndex = ajileSystem.GetProject().GetComponentIndexWithDeviceType(sequence.HardwareType());

            // stop any existing project from running on the device
            ajileSystem.GetDriver().StopSequence(componentIndex);

            // load the project to the device
            ajileSystem.GetDriver().LoadProject(project);
            ajileSystem.GetDriver().WaitForLoadComplete(-1);

            // run the project
            if (frameTime_ms >= 0)
                Console.WriteLine("Starting sequence " + sequence.ID().ToString() + " with frame rate " + frameTime_ms.ToString() + " and repeat count " + repeatCount.ToString());

            ajileSystem.GetDriver().StartSequence(sequenceID, componentIndex);

            // wait for the sequence to start
            Console.WriteLine("Waiting for sequence " + sequence.ID().ToString() + " to start");
            while (ajileSystem.GetDeviceState((byte)componentIndex).RunState() != RunState_e.RUN_STATE_RUNNING) ;

            if (repeatCount == 0)
            {
                Console.WriteLine("Sequence repeating forever. Press a key to stop the sequence");
                Console.ReadKey();
                ajileSystem.GetDriver().StopSequence(componentIndex);
            }

            Console.WriteLine("Waiting for the sequence to stop.");
            while (ajileSystem.GetDeviceState((byte)componentIndex).RunState() == RunState_e.RUN_STATE_RUNNING) ;

            return 0;
        }

        static void Main(string[] args)
        {
            RunExample(args);
        }
    }
}

File: examples/dmd_binary_testimages/matlab/dmd_binary_testimages.m
function dmd_binary_testimages(repeatCount, frameTime_ms, usbType)

    RunExample(repeatCount, frameTime_ms, usbType)
  
end

function [project] = CreateProject(sequenceID, sequenceRepeatCount, frameTime_ms, components)

    projectName = 'dmd_binary_testimages_example';
    filenameBase = '../../images/cat_';
    if frameTime_ms < 0
        frameTime_ms = 100;
    end
    numImages = 14;
    
    % create a new project
    project = py.ajiledriver.Project(projectName);
    project.SetComponents(components)
    
    % create the images
    for i = 1 : numImages
        filename = strcat(filenameBase, num2str(i), '.png');
        testImage = py.ajiledriver.Image(uint16(i));
        testImage.ReadFromFile(filename, py.ajiledriver.DMD_4500_DEVICE_TYPE);
        project.AddImage(testImage);
    end
    
    % create the sequence
    project.AddSequence(py.ajiledriver.Sequence(uint16(sequenceID), projectName, py.ajiledriver.DMD_4500_DEVICE_TYPE, py.ajiledriver.SEQ_TYPE_PRELOAD, uint16(sequenceRepeatCount)));

    % create a single sequence item, which all the frames will be added to
    project.AddSequenceItem(py.ajiledriver.SequenceItem(uint16(sequenceID), uint32(1)));

    for i = 1 : numImages
        frame = py.ajiledriver.Frame();
        frame.SetSequenceID(uint16(sequenceID));
        frame.SetImageID(uint16(i));
        frame.SetFrameTimeMSec(frameTime_ms);
        project.AddFrame(frame);
    end

end

function RunExample(repeatCount, frameTime_ms, usbType)

    % default connection settings
    ipAddress = '192.168.200.1';
    netmask = '255.255.255.0';
    gateway = '0.0.0.0';
    port = uint16(5005);
    if usbType == 2
        commInterface = py.ajiledriver.USB2_INTERFACE_TYPE;
    else
        commInterface = py.ajiledriver.USB3_INTERFACE_TYPE;
    end

    sequenceID = uint16(1);

    % connect to the device
    ajileSystem = py.ajiledriver.HostSystem();
    ajileSystem.SetConnectionSettingsStr(ipAddress, netmask, gateway, port);
    ajileSystem.SetCommunicationInterface(commInterface);
    if ajileSystem.StartSystem() ~= py.ajiledriver.ERROR_NONE
        'Error starting AjileSystem.'
        return;
    end
        
    % create the project
    project = CreateProject(sequenceID, repeatCount, frameTime_ms, ajileSystem.GetProject().Components());

    % get the first valid component index which will run the sequence
    result = cell(project.FindSequence(sequenceID));
    sequence = result{1};
    wasFound = result{2};
    if false(wasFound)
        'Sequence not found.'
        return;
    end
    componentIndex = ajileSystem.GetProject().GetComponentIndexWithDeviceType(sequence.HardwareType());

    % stop any existing project from running on the device
    ajileSystem.GetDriver().StopSequence(componentIndex);

    % load the project to the device
    ajileSystem.GetDriver().LoadProject(project);
    ajileSystem.GetDriver().WaitForLoadComplete(int32(-1));
            
    % run the project
    if frameTime_ms > 0
        sprintf('Starting sequence %d with frame rate %f and repeat count %d', sequence.ID(), frameTime_ms, repeatCount)
    end

    ajileSystem.GetDriver().StartSequence(sequence.ID(), componentIndex);

    % wait for the sequence to start
    sprintf('Waiting for sequence %d to start', sequence.ID())
    while ajileSystem.GetDeviceState(uint8(componentIndex)).RunState() ~= py.ajiledriver.RUN_STATE_RUNNING
        continue;
    end

    if repeatCount == 0
        'Sequence repeating forever. Press a key or click to stop the sequence'
        waitforbuttonpress;
        ajileSystem.GetDriver().StopSequence(uint8(componentIndex));
    end

    'Waiting for the sequence to stop.'
    while ajileSystem.GetDeviceState(uint8(componentIndex)).RunState() == py.ajiledriver.RUN_STATE_RUNNING
        continue;
    end

end
